tuf/signed/sign_test.go

```diff
@@ -264,9 +267,87 @@ func TestSignWithX509(t *testing.T) {
 	// RSAx509 key)
 	mockCryptoService := &StrictMockCryptoService{MockCryptoService{privKey}}
 	testData := data.Signed{}
-	err = Sign(mockCryptoService, &testData, tufRSAx509Key)
+	baseRole := data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		tufRSAx509Key,
+	)
+	err = Sign(mockCryptoService, &testData, baseRole)
 	assert.NoError(t, err)
 
 	assert.Len(t, testData.Signatures, 1)
 	assert.Equal(t, tufRSAx509Key.ID(), testData.Signatures[0].KeyID)
 }
+
+func TestSignWithOldKey(t *testing.T) {
+	cs := NewEd25519()
+	testData := data.Signed{}
+
+	key1, err := cs.Create(data.CanonicalRootRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	baseRole := data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		key1,
+	)
+	Sign(cs, &testData, baseRole)
+
+	key2, err := cs.Create(data.CanonicalRootRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	baseRole = data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		key2,
+	)
+	// should remove key1 sig even though it's valid. It no longer appears
+	// in the list of signing keys for the role
+	Sign(
+		cs,
+		&testData,
+		baseRole,
+	)
+
+	assert.Len(t, testData.Signatures, 1)
+	assert.Equal(t, key2.ID(), testData.Signatures[0].KeyID)
+}
+
+func TestSignWithOldSig(t *testing.T) {
+	cs := NewEd25519()
+	testData := data.Signed{}
+
+	key1, err := cs.Create(data.CanonicalRootRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	baseRole := data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		key1,
+	)
+	Sign(cs, &testData, baseRole)
+
+	// we need cs to "forget" key1 so we can't sign with it
+	cs = NewEd25519()
+	key2, err := cs.Create(data.CanonicalRootRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	// modify test data to invalidate key1 sig
+	testData.Signed = []byte{0xff}
+
+	baseRole = data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		key1,
+		key2,
+	)
+	// should remove key1 sig because it's out of date
+	Sign(
+		cs,
+		&testData,
+		baseRole,
+	)
+
+	assert.Len(t, testData.Signatures, 1)
+	assert.Equal(t, key2.ID(), testData.Signatures[0].KeyID)
+}
```

Comment:
Yes agreed that the role doesn't matter, I was just thinking of a use case since with delegations we may have many keys for the role but only sign with one at a time.

This is more of a nit since your test cases show the functionality, I figure it'd be nice to have for illustration purposes.